src/AppRun.c

```diff
@@ -25,182 +25,298 @@ THE SOFTWARE.
+        }
+    } while (exe[++n] != '\0');
+
+    // Create a pointer to each "substring" within `line`.
+    char* arg = exe;
+    // Counter for the number of arguments found.
+    int argcount = 0;
+
+    // Loop through each memory-adjacent string, counting them up.
+    while ((arg += (strlen(arg)+1)) && *arg != '\0') {
         argcount += 1;
+    }
 
-    // merge args
-    char*   outargptrs[argcount + argc + 1];
+    /* Merge the `Exec=` arguments with the `argv` arguments
+     * from the call to `main()`.
+     */
+
+    // Store the final list of arguments as an array of C-strings.
+    char* outargptrs[argcount + argc + 1];
+    // Store the name of the executable as the first argument.
     outargptrs[0] = exe;
-    int     outargindex = 1;
-    arg                 = exe;
-    int     argc_       = argc - 1;     // argv[0] is the filename
-    char**  argv_       = argv + 1;
+    // Keep track of which argument (index) we're storing.
+    int outargindex = 1;
+    /* Reset the `arg` pointer to the beginning of the sequence of
+     * memory-adjacent strings.
+     */
+    arg = exe;
+    // Skip the first argument, which is the filename, from the `main()` call.
+    int argc_ = argc - 1;
+    char** argv_ = argv + 1;
+    // Iterate through each argument from `Exec=`.
     while ((arg += (strlen(arg)+1)) && *arg) {
-        if (arg[0] == '%' || (arg[0] == '"' && arg[1] == '%')) {         // handle desktop file field codes
+        // If the current argument is a desktop file field code.
+        if (arg[0] == '%' || (arg[0] == '"' && arg[1] == '%')) {
+            // Get the actual code after the `%` or `"%`.
             char code = arg[arg[0] == '%' ? 1 : 2];
             switch(code) {
+                // For single file or URL args, merge in one of the `argv` args.
                 case 'f':
                 case 'u':
                     if (argc_ > 0) {
                         outargptrs[outargindex++] = *argv_++;
-                        argc_--;
+                        --argc_;
                     }
                     break;
+                /* For multiple file or URL args, merge in the rest of the
+                 * `argv` args.
+                 */
                 case 'F':
                 case 'U':
                     while (argc_ > 0) {
                         outargptrs[outargindex++] = *argv_++;
-                        argc_--;
+                        --argc_;
                     }
                     break;
+                /* The Icon (i), translated name (c), and desktop file location
+                * (k) keys are not supported. Throw a non-fatal error.
+                */
                 case 'i':
                 case 'c':
                 case 'k':
-                    fprintf(stderr, "WARNING: Desktop file field code %%%c is not currently supported\n", code);
+                    fprintf(stderr, "WARNING: Desktop field field %%%c is not currently supported.\n", code);
```

Comment:
"field field" -- that must be a mistake.

@probonopd prefers messages not to end with a stop.